### PR TITLE
GPU support for DynamicPruningExpression and InSubqueryExec

### PR DESCRIPTION
**Pull request by jlowe**
This adds GPU support for DynamicPruningExpression and InSubqueryExec which has been seen outside of the normal DPP filter expressions within a scan on some Spark platforms.  The implementations are relatively straightforward with the caveat that GpuInSubqueryExec needs to ensure the results are serialized to the executor since we do not participate in codegen like the CPU version does which the CPU relies on to get the results to the executor (i.e.: results are in the generated source code sent to the executors).

The support is only added for Spark 3.3+ on non-Databricks platforms.  Databricks has a different number of arguments for InSubqueryExec that would need to be investigated, and Spark versions prior to 3.3 also have differing arguments.  We don't see InSubqueryExec on any Spark platforms before Spark 3.3+, so not having a GPU replacement on these platforms should not be an issue in practice.

### CODE REVIEW

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
Manually tested that this PR resolves the fallback issues in the NDS benchmrak for DynamicPruningExpression and InSubqueryExec on AWS EMR 6.12.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDynamicPruningExpression.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import ai.rapids.cudf.ColumnVector
+
+import org.apache.spark.sql.catalyst.expressions.{DynamicPruning, Expression}
+
+case class GpuDynamicPruningExpression(child: Expression)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDynamicPruningExpression.scala:23]: Would we ever want this to return a scalar value? Like we do for `GpuKnownNotNull` and `GpuKnownFloatingPointNormalized`?

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/GpuInSubqueryExecSuite.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.nvidia.spark.rapids.{GpuOverrides, GpuTransitionOverrides, SparkQueryCompareTestSuite}
+
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.catalyst.expressions.{DynamicPruningExpression, ExprId}
+import org.apache.spark.sql.execution.{FilterExec, InSubqueryExec, LocalTableScanExec, ProjectExec, RowToColumnarExec, SparkPlan, SubqueryExec}
+
+class GpuInSubqueryExecSuite extends SparkQueryCompareTestSuite {
+  private def readToPhysicalPlan(df: DataFrame): SparkPlan = {
+    df.queryExecution.executedPlan.transformUp {
```
Comment (revans2) [inline tests/src/test/spark330/scala/org/apache/spark/sql/rapids/GpuInSubqueryExecSuite.scala:36]: Could we have some comments about what is happening here? Even though the code here is simple, it is not clear exactly what is happening or why.

**Review by revans2 — APPROVED** [review]
Really just some nits

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDynamicPruningExpression.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDynamicPruningExpression.scala:25]: Nice catch!  Fixed.

File: tests/src/test/spark330/scala/org/apache/spark/sql/rapids/GpuInSubqueryExecSuite.scala (hunk 2 above)
Comment (jlowe) [reply tests/src/test/spark330/scala/org/apache/spark/sql/rapids/GpuInSubqueryExecSuite.scala:49]: Added some comments here.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

**Review by revans2 — DISMISSED** [review]

**Comment by jlowe** [comment]
build

**Review by revans2 — DISMISSED** [review]

**Comment by jlowe** [comment]
Seeing some odd results when benchmarking this at scale, marking this as draft while it is investigated.

**Comment by jlowe** [comment]
The runs at scale are quite noisy.  This seems to significantly help NDS query 37 and 82, but the results for the remaining queries that had DynamicPruningExpression and InSubqueryExec fallbacks on AWS EMR 6.12 were too noisy to tell.

**Comment by jlowe** [comment]
build

**Comment by jlowe** [comment]
build

**Review by revans2 — APPROVED** [review]

**Pull request merged**